### DISCUSSION TITLE
Crop using mask

### DISCUSSION THREAD

**Discussion by jukkakahkonen** [ImageMagick/ImageMagick#7104] [Help]
### ImageMagick version

7.1.0-55

### Operating system, version and so on

Windows 11 65 bit

### Description

Crop image from command line. 
![BTI-351321](https://github.com/ImageMagick/ImageMagick/assets/78900138/dba0a658-84c4-448f-af41-b48696ae52fb)

This is not only this time. Many times working with similar problem. 

Final result: Cropped image, resized to exact dimensions 114x91 mm. Next step: make one image using several this type images. 

I test many way crop this image. 
`convert input.jpg -fuzz 10% -fill none -draw "color 0,0 floodfill" -channel alpha -blur 0x2 -level 50%,100% +channel mask.png
`
Of course this not work reason image border (white) and image (aluminium) color is quite near. Also all this type jinx touch all areas in image. Not only border. 

Reason all images are similar it is clever way make mask **manually**. So I take one image and using "photoshop" I make mask. 

![mask](https://github.com/ImageMagick/ImageMagick/assets/78900138/886fefe4-d153-478f-a0d2-e5f1ae1337cd)

Ok, this is maybe most clever way. 

`convert input.jpg mask.png -compose copy_opacity -composite output.png`

Problem is always: Crop work. But need just it "crop to mask". If I make this using "photoshop" manyally, simply mask using mouse and then "crop to mask". 

So, look this "crop to mask" or similar function is very complex in imagemagic. Maybe this "make mask manually" is good idea reason images are similar. But how, I cannot found. So- how "crop to mask"? All ideas I test are stupid, crop, but leave transparent area. 

-- 
Next question is, "How I can resize this image to right size, 114x91 mm".

**Comment by snibgo** [comment]
> Crop image from command line.

Your file includes a clipping path. Do you want to use that? I will assume you don't.

Trimming with a fuzz of 10% works well. I have named your input file as `jukka.jpg`.
```
magick jukka.jpg -fuzz 10% -trim +repage x.png
```
The result is 992x795 pixels. That ratio is close your your desired "exact dimensions 114x91 mm".

If we want 992 pixels to be 114 mm, this is 992/114 pixels per mm, or 992/11.4 pixels per cm, 87.02 pixels per cm. So we can change the density:
```
magick jukka.jpg -fuzz 10% -trim +repage -density 87.02 -units pixelspercentimeter x.png
```
The result is still 992x795 pixels, physical dimensions:
```
  Print size: 11.4023x9.13793
  Units: PixelsPerCentimeter
```

I don't understand what you want with a manual mask. But if you have one, you can find the crop parameters with:
```
magick mask.png -format "%@" info:

**Comment by jukkakahkonen** [comment]
Very good, thanks. Problem is, only manual for Imagemagick is Chatgpt. I try google imagemagick manual many hours- nothing. Now I try study "montage". Same problem- no manual. Is it any manual for Imagemagick? I try search "Imagemagick -montage manual" etc, but nothing. Is it any hidden place for manual? Or any excelent "tip how to search using google"? (Eg. "montage -help" is not good, it put only command list, but eg. "montage -help tile" etc not work...) 

So, really need good manual...

**Comment by jukkakahkonen** [comment]
Ok, now crop proceeded and ok. I try make several combinations using "components". 

![frame](https://github.com/ImageMagick/ImageMagick/assets/78900138/5133631e-4604-4479-a5f0-2c9041584bbf)
![lid](https://github.com/ImageMagick/ImageMagick/assets/78900138/71f87358-430c-448f-8f5f-5146a3b9a5a3)
![speaker](https://github.com/ImageMagick/ImageMagick/assets/78900138/e58de0ef-9fd0-40b7-aa27-20f1bd31b25e)
![4btn](https://github.com/ImageMagick/ImageMagick/assets/78900138/bafea01a-f4b5-415e-ab46-cab15083d74e)

Original images are quite high-resolution png. Adjusted to right size, it is, images are 600 dpi. Dimensionally images are realistic etc. 
Using components final result will be example: 
![result](https://github.com/ImageMagick/ImageMagick/assets/78900138/958e9403-5382-4f8f-a7ed-02c9b5b3c1ad)

Combinations found about 100 different. Different components quantity is about 20. So, lid + frame. Then add lid+frame-combinations 1, 2, 3 or even 4. Then put "components" inside frames exact grid. (Here is only two example, "4button" and "speaker". Other components are "8button", "camera" etc etc.... and all are same size.)

Ok: 
`$kuva1 = "frame.png"
$kuva2 = "lid.png"
montage $kuva2 $kuva1 -geometry +0+0 -tile 2x1 BTI-350231_complete.png
`
This make first piece:
![BTI-350231_complete](https://github.com/ImageMagick/ImageMagick/assets/78900138/f863895f-3126-4eff-b238-b536550ac771)

QUESTION: 
- How I can put other "components"? 
- As you see, "speaker.png" and "4button.png" found area. Hiddenable area. So: background must be frame. Front of it must be "component", eg. "speaker.png". And front of it must be "lid.png". Is this possible or maybe I must crop it extra area away?

**Comment by snibgo** [comment]
Montage is documented at https://imagemagick.org/script/montage.php

For general ImageMagick documentation, I suggest starting with:
https://imagemagick.org/index.php
http://imagemagick.org/Usage/
http://imagemagick.org/Usage/basics/
http://imagemagick.org/Usage/windows/
https://imagemagick.org/script/command-line-options.php
https://imagemagick.org/script/defines.php

Most IM documentation examples are for bash. Examples on my pages at http://im.snibgo.com/ are mostly for Windows BAT scripts.

Your components can be combined using `+append`, `-append` and `composite`. For example (Windows BAT syntax):
```
magick ^
  ( ^
    jukka1.png ^
    ( jukka4.png jukka4.png jukka4.png -append +repage ) ^
    -gravity Center ^
    -compose Over -composite ^
    jukka2.png ^
    +swap +append +repage ^
  ) ^
  ( ^
    jukka1.png ^
    ( jukka4.png jukka3.png jukka3.png -append +repage ) ^
    -gravity Center ^
    -compose Over -composite ^
    jukka2.png ^
    +swap +append +repage ^
  ) ^
  +append +repage ^
  x.png
```
The result, x.png, is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/00cb59ad-7e24-42d2-a271-d4e7c3c5a152)

**Reply by jukkakahkonen** [reply]
Okayyy :D. Thank you, and really, much! I will build this with powershell script. (And maybe later I will make automatic tool using php etc etc :D. ) 

Now I must say "wow!". Excelent!

**Comment by jukkakahkonen** [comment]
Hm... look it is not easy way translate this .bat to powershell .ps1-format. Also I try use variables reason then it will more flexible... 

`$kuva1 = "jukka1.png"
$kuva2 = "jukka2.png"
$kuva3 = "jukka3.png"
$kuva4 = "jukka4.png"

magick `
  ( `
    $kuva1 `
    ( $kuva4 $kuva4 $kuva4 -append +repage ) `
    -gravity Center `
    -compose Over -composite `
    $kuva2 `
    +swap +append +repage `
  ) `
  ( `
    $kuva1 `
    ( $kuva4 $kuva3 $kuva3 -append +repage ) `
    -gravity Center `
    -compose Over -composite `
    $kuva2 `
    +swap +append +repage `
  ) `
  +append +repage `
  x.png
`

Unexpected token '(' in expression or statement.
At ...........kokoa.ps1:9 char:11
+     $kuva1 `
+           ~
Missing closing ')' in expression.

Etc etc. 

`$kuva1 = "jukka1.png"
$kuva2 = "jukka2.png"
$kuva3 = "jukka3.png"
$kuva4 = "jukka4.png"

magick `
  ( $kuva1, ($kuva4, $kuva4, $kuva4 -join " -append +repage") -join " ", $kuva2 -join " +swap +append +repage") `
  ( $kuva1, ($kuva4, $kuva3, $kuva3 -join " -append +repage") -join " ", $kuva2 -join " +swap +append +repage") `
  +append +repage `
  x.png
`

Also this not work. Maybe problem is my skills- powershell scripts are more flexible so this is reason I try use it instead bat. But translate from bat to .ps1 is not so easy. Also it is possible problem found wrong use variables.

**Reply by jukkakahkonen** [reply]
Also this not work...  
`# Kokoamiskomento
$montageCommand = "magick ( `"$kuva1`" ( `"$kuva4`" `"$kuva4`" `"$kuva4`" -append +repage ) -gravity Center -compose Over -composite `"$kuva2`" +swap +append +repage ) ( `"$kuva1`" ( `"$kuva4`" `"$kuva3`" `"$kuva3`" -append +repage ) -gravity Center -compose Over -composite `"$kuva2`" +swap +append +repage ) +append +repage `"$tulos`""

# Suoritetaan kokoamiskomento
Invoke-Expression $montageCommand
`

Look problem is, using powershell this is very difficult to make. So maybe only way is use .bat? (Reason why I use Powershell is simply- it script language is near real programming language. So it is much easier use it eg. webpage etc. ) I really cannot understand why this not work- in nutcell, easy but not easy...

**Comment by snibgo** [comment]
It can be done in Powershell. If you want to use Powershell but don't know how to, then I suggest you learn it.

**Reply by jukkakahkonen** [reply]
Hmm... studying and testing this two days. Googling and using chatgpt. Chatgpt is not good in this type studying, it recommend "change this " to '  or ^" etc, without any logic. Also powershell manuals does not tell how to solve this type problem. Look bat-scripts and powershell-scripts are really different, so, philosophy is much, much different. Final result look, this type work is not possible with powershell-script? Or, maybe problem is, it is not possible use variables in this type work?

**Comment by snibgo** [comment]
> Look bat-scripts and powershell-scripts are really different, so, philosophy is much, much different. 

Yes.

If you don't know Powershell, why do you want to use Powershell?

Here is a Powershell script for your command:
```
$kuva1 = "jukka1.png"
$kuva2 = "jukka2.png"
$kuva3 = "jukka3.png"
$kuva4 = "jukka4.png"

$cmdargs = -split @"
(
$kuva1 ( $kuva4 $kuva4 $kuva4 -append +repage )
-gravity Center -compose Over -composite
$kuva2 +swap +append +repage
) (
$kuva1 ( $kuva4 $kuva3 $kuva3 -append +repage )
-gravity Center -compose Over -composite
$kuva2 +swap +append +repage
) +append +repage
x.png
"@

magick $cmdargs
```
